File: commcare-core/src/test/java/org/commcare/xml/ParserTestUtils.java
package org.commcare.xml;

import org.javarosa.xml.ElementParser;
import org.javarosa.xml.PlatformXmlParser;

import java.io.ByteArrayInputStream;
import java.lang.reflect.Constructor;
import java.util.function.Function;

/**
 * Helper functions for building parsers
 */
public class ParserTestUtils {

    public static <T extends CommCareElementParser> T buildParser(String xml, Class<T> parserClass) {
        return buildParser(xml, (xmlParser) -> {
            try {
                Constructor<T> constructor = parserClass.getConstructor(PlatformXmlParser.class);
                return constructor.newInstance(xmlParser);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        });
    }

    public static <T extends CommCareElementParser> T buildParser(String xml, Function<PlatformXmlParser, T> builder) {
        try {
            ByteArrayInputStream inputStream = new ByteArrayInputStream(xml.getBytes("UTF-8"));
            PlatformXmlParser parser = ElementParser.instantiateParser(inputStream);
            return builder.apply(parser);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: commcare-core/src/test/java/org/commcare/xml/ParserTestUtils.kt
package org.commcare.xml

import org.javarosa.xml.ElementParser
import org.javarosa.xml.PlatformXmlParser
import java.io.ByteArrayInputStream
import java.util.function.Function

/**
 * Helper functions for building parsers
 */
class ParserTestUtils {
    companion object {
        @JvmStatic
        fun <T : CommCareElementParser<*>> buildParser(xml: String, parserClass: Class<T>): T {
            return buildParser(xml, Function { xmlParser ->
                try {
                    val constructor = parserClass.getConstructor(PlatformXmlParser::class.java)
                    constructor.newInstance(xmlParser)
                } catch (e: Exception) {
                    throw RuntimeException(e)
                }
            })
        }

        @JvmStatic
        fun <T : CommCareElementParser<*>> buildParser(xml: String, builder: Function<PlatformXmlParser, T>): T {
            try {
                val inputStream = ByteArrayInputStream(xml.toByteArray(Charsets.UTF_8))
                val parser = ElementParser.instantiateParser(inputStream)
                return builder.apply(parser)
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
        }
    }
}
